metadata-ingestion-modules/airflow-plugin/src/datahub_airflow_plugin/datahub_listener.py

```diff
@@ -0,0 +1,411 @@
+import copy
+import functools
+import logging
+import threading
+from typing import TYPE_CHECKING, Callable, Dict, Optional
+
+import datahub.emitter.mce_builder as builder
+from airflow.listeners import hookimpl
+from datahub.api.entities.datajob import DataJob
+from datahub.api.entities.dataprocess.dataprocess_instance import InstanceRunResult
+from datahub.emitter.rest_emitter import DatahubRestEmitter
+from datahub.metadata.schema_classes import (
+    FineGrainedLineageClass,
+    FineGrainedLineageDownstreamTypeClass,
+    FineGrainedLineageUpstreamTypeClass,
+)
+from datahub.utilities.sqlglot_lineage import SqlParsingResult
+from openlineage.airflow.listener import TaskHolder
+from openlineage.airflow.utils import redact_with_exclusions
+from openlineage.client.serde import Serde
+
+from datahub_airflow_plugin._airflow_shims import (
+    Operator,
+    get_task_inlets,
+    get_task_outlets,
+)
+from datahub_airflow_plugin._config import DatahubLineageConfig, get_lineage_config
+from datahub_airflow_plugin._datahub_ol_adapter import translate_ol_to_datahub_urn
+from datahub_airflow_plugin._extractors import ExtractorManager
+from datahub_airflow_plugin.client.airflow_generator import AirflowGenerator
+from datahub_airflow_plugin.entities import _Entity
+
+if TYPE_CHECKING:
+    from airflow.models import DagRun, TaskInstance
+    from datahub.ingestion.graph.client import DataHubGraph
+    from sqlalchemy.orm import Session
+
+logger = logging.getLogger(__name__)
+
+_airflow_listener_initialized = False
+_airflow_listener: Optional["DataHubListener"] = None
+_RUN_IN_THREAD = True
+
+
+def get_airflow_plugin_listener() -> Optional["DataHubListener"]:
+    # Using globals instead of functools.lru_cache to make testing easier.
+    global _airflow_listener_initialized
+    global _airflow_listener
+
+    if not _airflow_listener_initialized:
+        _airflow_listener_initialized = True
+
+        plugin_config = get_lineage_config()
+
+        if plugin_config.enabled:
+            _airflow_listener = DataHubListener(config=plugin_config)
+
+            if plugin_config.disable_openlineage_plugin:
+                # Deactivate the OpenLineagePlugin listener to avoid conflicts.
+                from openlineage.airflow.plugin import OpenLineagePlugin
+
+                OpenLineagePlugin.listeners = []
+
+    return _airflow_listener
+
+
+def run_in_thread(f):
+    @functools.wraps(f)
+    def wrapper(*args, **kwargs):
+        try:
+            if _RUN_IN_THREAD:
+                thread = threading.Thread(
+                    target=f, args=args, kwargs=kwargs, daemon=True
+                )
+                thread.start()
+
+                thread.join(timeout=10)
```

Comment:
What does this give us since we're joining immediately after anyway?

Reply:
it gives us a mechanism to kill the extractor after 10 seconds

this is a poor-man's timeout - added a comment